### PR TITLE
WIP - start of README overhaul

### PR DESCRIPTION
**Pull request by lholmquist**
connects to #126

### CODE REVIEW

**Comment by lholmquist** [comment]
rebased with latest master

File: README.md [hunk 1]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
+### Resource Fragments
 
-Notice that there is a `.nodeshift` directory in the project, this will hold application specific resource files for creating Routes/Serivces/DeploymentConfigs
+OpenShift resource fragments are user provided YAML files that can be enriched with metadata, labels and more by nodeshift.
 
-Before running the nodeshift command, we first need to login to our Openshift Cluster running on Minishift using the `oc login` command
+Each resource gets its own file, which contains some skeleton of a resource description. Nodeshift will enrich it and then combine all the resources into a single openshift.yml and openshift.json(located in ./tmp/nodeshift/resource/).
 
-`oc login -u developer` for example.
+The resource objects `Kind`, if not given, will be extracted from the filename.
```
Comment (lance) [inline README.md]: possessive: "object's"

File: README.md [hunk 2]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
-* OC command line tools
+**deploy** - does everything, the archiving of code and everything that apply-resource does
 
-These are also part of the minishift install, and [step 2 of this section](https://docs.openshift.org/latest/minishift/getting-started/quickstart.html#starting-minishift) tells you how to put them on your path.
+**undeploy** - removes things that were deployed with the apply-resource command
 
-This guide also assumes you have a modern version(4+) of Node.js and npm installed.
 
-#### Example
+### .Nodeshift Directory
```
Comment (lance) [inline README.md]: It's not totally clear from this description that it's the user's responsibility to supply the resource files in the `.nodeshift` directory. What is the minimum set of resource files required for a project?

File: README.md [hunk 3]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
+                  - name: REALM
+                    value: master
 
+To set that using nodeshift, use the `-d` option with a KEY=VALUE, like this:
 
-Or just calling nodeshift directly if you have it installed globally
+    nodeshift -d SSO_AUTH_SERVER_URL=https://sercure-url
 
-    $ nodeshift
+__note that we left off the "${}" on the key,  nodeshift knows to search for a key with ${} added back on__
```
Comment (lance) [inline README.md]: Not sure this is entirely necessary.

File: README.md [hunk 4]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
+    })
 
-**apply-resource** - does the resource goal and will also push that to your running cluster
+_please note: Currently, once a route, service, deployment config, build config, and imagestream config are created, those are re-used. The only thing that changes from deployment to deployment is the source code.  For application resources, you can update them by undeploying and then deploying again.  BuildConfigs and Imagestreams can be re-created using the --build.recreate flag_
 
-**deploy** - does everything, the archiving of code and everything that apply-resource does
+## Advanced Options
 
-**undeploy** - removes things that were deployed with the resource command
+While nodeshift tries to be very opinionated during the deployment process, but there are options available to pass to the cli or the API.
```
Comment (lance) [inline README.md]: This sentence is awkward. How about

While nodeshift is very opinionated about deployment parameters, both the CLI and the API accept options that allow you to customize nodeshift's behavior.

File: README.md [hunk 5]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
 
-**build** - archives the code, creates build config and imagestream and pushes the binary to the cluster
+#### version
+Outputs the current version of nodeshift
 
-#### Template Parameters
+#### projectLocation
+Changes the default location of where to look for your project. Defaults to your current working directory(CWD)
```
Comment (lance) [inline README.md]: s/of where to look//

File: README.md [hunk 6]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
 
-Some templates might need to have a value set at "run time".  For example, in the template below, we have the `${SSO_AUTH_SERVER_URL}` parameter:
+#### configLocation
+Changes the default location of where to look for a configfile. Defaults to ~/.kube/
```
Comment (lance) [inline README.md]: s/of where to look for a configfile/of the config file/

Comment (lance) [reply README.md]: What goes in the config file?

File: README.md [hunk 7]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
-          template:
-            spec:
-              containers:
-                - env:
-                  - name: SSO_AUTH_SERVER_URL
-                    value: "${SSO_AUTH_SERVER_URL}"
-                  - name: REALM
-                    value: master
+#### osc.strictSSL
+Setting to pass to the [Openshift Rest Client](https://www.npmjs.com/package/openshift-rest-client) for SSL use.  To allow using a self-signed cert, set to false
```
Comment (lance) [inline README.md]: Maybe here and in `osl.tryServiceAccount` the wording could be a little clearer. It took me a few reads to understand what you mean. How about "This option is passed through to the ...".

File: README.md [hunk 8]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
 
-To set that using nodeshift, use the `-d` option with a KEY=VALUE, like this:
+#### osl.tryServiceAccount
```
Comment (lance) [inline README.md]: Why is this prefix `osl` and the other is `osc`. Should we be consistent?

File: README.md [hunk 9]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
+Setting to pass to the [Openshift Config Loader](https://www.npmjs.com/package/openshift-config-loader). Set to false to by-pass service account lookup or use the KUBERNETES_AUTH_TRYSERVICEACCOUNT environment variable
 
-    nodeshift -d SSO_AUTH_SERVER_URL=https://sercure-url
+#### nodeVersion
+Specify the version of Node.js to use for the deployed application. defaults to latest.  These version tags corespond to the docker hub tags of the [bucharest-gold s2i images](https://hub.docker.com/r/bucharestgold/centos7-s2i-nodejs/tags/)
 
-__note that we left off the "${}" on the key,  nodeshift knows to search for a key with ${} added back on__
+#### quiet
```
Comment (lance) [inline README.md]: Actually, this is an environment variable. I'm not sure why I did it this way (and maybe it should change to be a CLI option). In any case, to suppress `INFO` and `TRACE` lines, you have to `NODESHIFT_QUIET=true nodeshift`.

File: README.md [hunk 10]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
+supress INFO and TRACE lines from output logs.
 
-For more on writing openshift templates, [see here](https://docs.openshift.org/latest/dev_guide/templates.html#writing-templates)
+#### build.recreate
+Flag to recreate a buildConfig or Imagestream.  Defaults to false. Choices are "buildConfig", "imageStream", false, true
```
Comment (lance) [inline README.md]: If `true` are both the `BuildConfig` and `ImageStream` recreated? Also, capitalize `BuildConfig` in the first sentence.

File: README.md (hunk 8 above)
Comment (lholmquist) [reply README.md]: one is for openshift-rest-client(osc)  and the other is for openshift-config-loader(osl)

really it should probably be orc and ocl

File: README.md (hunk 6 above)
Comment (lholmquist) [reply README.md]: tbh,  i don't know if we should really even let the user do this.  All that config stuff is created and managed by the oc commands(and kubectl also),

Comment (lance) [reply README.md]: I agree. I don't think it is something that needs to be exposed.

File: README.md (hunk 8 above)
Comment (lance) [reply README.md]: I guess I just wonder if it really matters for the end user. They don't care whether it's the rest client or the config loader, and they don't need to know. So, having a prefix that differs between these options doesn't seem to buy anything. Could they not both just be `os` for OpenShift?

File: README.md [hunk 11]
```diff
@@ -1,12 +1,15 @@
...
 
-Nodeshift is a Plugin for running openshift deployments for node.js projects
+Nodeshift is an opinionated command line application and programmable API that you can use to deploy Node.js projects to OpenShift. It's currently under heavy development and APIs are still shifting a bit. However, in spite of its beta nature, you can use nodeshift today to deploy Node.js applications to OpenShift.
 
-_Disclaimer: Nodeshift is extremely experimental at the moment.  This means it is under heavy development, api's could change at any time until a 1.0.0 release_
+## Prerequisites
 
-### Install
+* Node.js - version 8.x or greater
+* npm - version 4.x or greater
```
[the remark this answers is not in the document]
Comment (lholmquist) [reply README.md]: probably not

File: README.md [hunk 12]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
-### Example Usage
-
-#### Prerequisites
+## Core Concepts
 
-* Minishift
+### Commands & Goals
 
-For this example, we will run against an Openshift Cluster running locally using Minishift.
+By default, if you run just `nodeshift`, it will run the `deploy` goal, which runs everything.
```
[the remark this answers is not in the document]
Comment (lholmquist) [reply README.md]: +1

File: README.md [hunk 13]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
-_I installed using Homebrew and use Virtual Box as my hypervisor provider, so my command to initially start minishift was something like this: `minishift start --vm-driver=virtualbox`_
+**apply-resource** - does the resource goal and will also push that to your running cluster
 
+**build** - archives the code, creates build config and imagestream and pushes the binary to the cluster
 
-* OC command line tools
+**deploy** - does everything, the archiving of code and everything that apply-resource does
 
-These are also part of the minishift install, and [step 2 of this section](https://docs.openshift.org/latest/minishift/getting-started/quickstart.html#starting-minishift) tells you how to put them on your path.
+**undeploy** - removes things that were deployed with the apply-resource command
```
[the remark this answers is not in the document]
Comment (lholmquist) [reply README.md]: s/things/resources

File: README.md [hunk 14]
```diff
@@ -21,147 +24,131 @@ or to use in an npm script
...
+The `.nodeshift` directory is responsible for holding your resource fragements.  These are `.yml` files that describe you service, deployments, routes, etc.
 
-_From this point on, all commands will assume you are in the directory of this source code_
+### Resource Fragments
 
-Notice that there is a `.nodeshift` directory in the project, this will hold application specific resource files for creating Routes/Serivces/DeploymentConfigs
+OpenShift resource fragments are user provided YAML files that can be enriched with metadata, labels and more by nodeshift.
 
-Before running the nodeshift command, we first need to login to our Openshift Cluster running on Minishift using the `oc login` command
+Each resource gets its own file, which contains some skeleton of a resource description. Nodeshift will enrich it and then combine all the resources into a single openshift.yml and openshift.json(located in ./tmp/nodeshift/resource/).
```
[the remark this answers is not in the document]
Comment (lholmquist) [reply README.md]: I think the Enricher heading sort of says this

File: README.md (hunk 4 above)
Comment (lholmquist) [reply README.md]: english is hard :)

File: README.md (hunk 6 above)
Comment (lholmquist) [reply README.md]: ok,  removing this

File: README.md (hunk 8 above)
Comment (lholmquist) [reply README.md]: i think that makes sense.  I know the boosters are using that `osc.strictSSL` flag, so we need to remember to update that :)

File: README.md (hunk 9 above)
Comment (lholmquist) [reply README.md]: I thought it was an option,  https://github.com/bucharest-gold/nodeshift/blob/master/bin/nodeshift#L63

File: README.md (hunk 10 above)
Comment (lholmquist) [reply README.md]: correct, a value of true will do both,  i'll make that more clear

File: README.md (hunk 8 above)
Comment (lholmquist) [reply README.md]: i'm thinking, does it even need to be `os.something` ,  if we are hiding the rest-client and configloader from the user,  then why not just make the flags `--strictSSL` and `--tryServiceAccount`

**Comment by lholmquist** [comment]
Thanks for the review,  i hate writing readme's  :)   anyway.  i think i got all the things.

I've removed the configLocation flag and changed the osl.tryServiceAccount and osc.strictSSL flags to just be `tryServiceAccount` and `strictSSL`

**Comment by lholmquist** [comment]
ok, now i think i fixed everything wrt the osl and osc stuff.  forgot about the public api part

File: README.md (hunk 9 above)
Comment (lance) [reply README.md]: You're right. I will fix it. https://github.com/bucharest-gold/nodeshift/issues/154

File: README.md (hunk 6 above)
Comment (jamesfalkner) [reply README.md]: I went to upgrade to latest nodeshift in [a project that runs in a Katacoda environment](https://learn.openshift.com/middleware/rhoar-getting-started-nodejs/), where the configuration file for the `oc` command is _not_ at `~/.kube/config`. It's actually at `/openshift.local.config/master/admin.kubeconfig`, and I was using the `configLocation` in nodeshift to point at it. I'm not an expert at kubernetes config locations and don't know if this is a bad practice, but removing this prevents my project from using latest nodeshift...

**Comment by lholmquist** [comment]
could you create a new issue so we can discuss there,  I can't seem to find
this comment on the closed PR

**Pull request merged**